SEED-platform/seed
# refactor: add media api for serving protected user files

#### Any background context you want to provide?

#### What's this PR do?
Adds permissioning in front of serving media files (ie user uploaded files).
More specifically, creates `/api/v3/media/<filepath>` (and sets it as the MEDIA_URL) which verifies the requesting user has access to the file, before instructing Nginx to serve the file using the X-Accel-Redirect header. If the user doesn't have access or the file doesn't exist, it 404s.

#### How should this be manually tested?
If using docker-compose setup, you must be using the non-dev version (ie docker-compose.local). This is because this update requires Nginx to serve the files.
Create two separate orgs with different user accounts
Try to access different media files while logged in as different users (or not logged in). Media file locations:
- /api/v3/media/uploads
    - (used for most property imports)
    - create by importing from a csv
- /api/v3/media/buildingsync_files/
    - (used for bsync imports)
    - create by importing bsync file
- /api/v3/media/analysis_input_files/<analysis id>/
    - (where input files for an analysis are stored)
    - create by running bsyncr analysis. (for refresher, [ see this PR](https://github.com/SEED-platform/seed/pull/2547))
- /api/v3/media/analysis_output_files/
    - (where analysis output files are stored)
    - create by running bsyncr analysis

Try to access `<host>/media/<some file>` directly and it should fail (/media is no longer the MEDIA_URL). Try to access `<host>/protected/<some file>` and it should fail (the protected location is _internal_ to Nginx, meaning it's only accessible when not requested externally, ie in client request)

Try to break it by using it in unexpected ways...

#### What are the relevant tickets?
#2626 

#### Screenshots (if appropriate)
